### DISCUSSION TITLE
How does `torchio.transforms.RescaleIntensity`'s in_min_max argument works?

### DISCUSSION THREAD

**Discussion by gRox167** [TorchIO-project/torchio#579] [Q&A]
`torchio.transforms.RescaleIntensity`'s in_min_max argument is not working properly or maybe I misunderstood the document.

**Expected behavior**
In the document, it is said that "Range (mmin,mmax) of input intensities that will be mapped to (nmin,nmax). If None, the minimum and maximum input intensities will be used." I thought it works like the function `torch.clamp` or `np.clip`, which will clip the in_min and in_max.
**Actual behavior**
In your source code, this step is done by 
```python
array -= in_min
in_range = in_max - in_min
array /= in_range
```
I think this just scale down the distribution of image intensity, but not clip off the value outside the range we set.

**Comment by romainVala** [comment]
Hello,

no there is a clip first : https://github.com/fepegar/torchio/blob/c95410bc1eeb8066907a3ba0eb152d75afdffd7d/torchio/transforms/preprocessing/intensity/rescale.py#L81
and then the rescale

**Comment by fepegar** [comment]
Hi, @gRox167. I'm not sure what the question is.

The clipping happens only if the `percentiles` kwarg is used (to remove potential outliers and make the transform more robust). Otherwise there's no clipping, just a linear transform to change the range of intensity values. You can take a look at the [`scikit-learn` docs for `rescale_intensity`](https://scikit-image.org/docs/dev/api/skimage.exposure.html#skimage.exposure.rescale_intensity), it might help.

**Reply by gRox167** [reply]
Thanks for your reply! But I'm still a little confused about the behavior of in_min_max in Torchio, for it's different from scikit-learn.
In `rescale_intensity` the document mentioned that if the min/max value of `in_range` is more/less than the min/max image    intensity, then the intensity levels are clipped
```
>>>image = np.array([51, 102, 153], dtype=np.uint8)
>>>image_float = 1.0 * image
>>>rescale_intensity(image_float, in_range=(0, 102))
array([0.5, 1. , 1. ])
```
The expected behavior of `in_min_max` is to clip off the outlier that smaller or larger than in_min or in_max, as mentioned in scikit-learn's `rescale_intensity`. So I wonder maybe the document is a little confuse for the novice user :), as it works differently from scikit's `rescale_intensity`.

**Reply by fepegar** [reply]
Well, I suppose the implementation in TorchIO is different, then. Do you think the skimage one makes more sense?

> The expected behavior of in_min_max is to clip off the outlier that smaller or larger than in_min or in_max, as mentioned in scikit-learn's rescale_intensity

That's not really true, the expected behaviour is not to clip. That's what `percentiles` does.

Let's see if I can upload a photo from my phone...

**Reply by fepegar** [reply]
So here `in1, in2 = in_min_max` and `out1, out2 = out_min_max`.

![IMG_20210621_171325](https://user-images.githubusercontent.com/12688084/122794580-519b2380-d2b4-11eb-85f8-d5a92b04f6c1.jpg)

Before that linear mapping, the array is clipped using the given percentiles to remove outliers.

**Reply by gRox167** [reply]
I guess I get it. But in TorchIO's code, the `in_min_max` only clip off the lower boundary limited by `in_min` and divide all intensity by `in_max-in_min`. Is there an inconsistency in its logic, why not clip off the upper bound like scikit? I know `percentiles` done the clip-off work, and `in_min_max` also do clip-off work in scikit style `rescale_intensity`, these 2 clip off are not compatible. So I guess it's better to let users choose which way to go.

And also your ways of implementation will cause the output result doesn't always fall into the range of (out_min,out_max). 
```
array -= in_min
in_range = in_max - in_min
array /= in_range
```
In this code snippet, if `in_max` is smaller than the original max intensity, the `array`'s range is not (0,1), max value will larger than 1, which will lead to a larger output value in the later step.

Also in the photo, if the whole x-axis represents the total intensity distribution such as (0,255), the $in_1$, $in_2$ limited the input of intensity to such as (50,200), then rescale voxel intensity between (50,200) to $out_1$, $out_2$. Does the photo represent the same behavior as a scikit style `rescale_intensity`. 
![image](https://user-images.githubusercontent.com/40270265/122876587-44d80780-d368-11eb-874c-3565e967b8e5.png)

Thanks for your kindly answer.
